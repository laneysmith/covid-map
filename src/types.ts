export const COLORS = [
    '#DEEDCF',
    '#B8DEAA',
    '#88CE86',
    '#62BE73',
    '#3FAC6C',
    '#1D9A6C',
    '#188977',
    '#137177',
    '#0E4D64',
    '#0A2F51',
];

export enum Variable {
    CASES = 'cases',
    DEATHS = 'deaths',
}

export type ColorScales = { [key in Variable]: [] };

export type FipsStats = { [key in Variable]: number }

export type ResponseData = { data: object, maxCases: number, maxDeaths: number }
