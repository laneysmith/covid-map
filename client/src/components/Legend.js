import React, { useMemo } from 'react';

import { CASES, DEATHS } from '../constants';
import { formatNumber, chunkArray } from '../utils';

const Legend = ({ selected, onChangeVariable, loaded, colorScales }) => {
  const handleChangeVariable = (e) => onChangeVariable(e.target.value);
  const scale = colorScales[selected];
  const chunkedScale = useMemo(() => {
    const chunked = scale.length > 0 ? chunkArray(scale) : null;
    return chunked;
  }, [scale]);

  return (
    <div id="state-legend" className="legend">
      <div className="legend-controls">
        <label htmlFor={CASES}>
          <input
            type="radio"
            id={CASES}
            name="variable"
            value={CASES}
            checked={selected === CASES}
            onChange={handleChangeVariable}
            disabled={!loaded}
          />{' '}
          Cases
        </label>
        <label htmlFor={DEATHS}>
          <input
            type="radio"
            id={DEATHS}
            name="variable"
            value={DEATHS}
            checked={selected === DEATHS}
            onChange={handleChangeVariable}
            disabled={!loaded}
          />{' '}
          Deaths
        </label>
      </div>
      <div className="legend-scale">
        {chunkedScale &&
          chunkedScale.map((step, index) => {
            const [value, color] = step;
            return (
              <div key={`scale-step-${index}`} id={`scale-step-${index}`}>
                <span style={{ backgroundColor: `${color}` }}></span> {formatNumber(value)}
              </div>
            );
          })}
      </div>
    </div>
  );
};

export default Legend;
