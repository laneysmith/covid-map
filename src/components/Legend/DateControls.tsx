import React, { HTMLAttributes } from 'react';

import { ReactComponent as PauseIcon } from './pause.svg';
import { ReactComponent as PlayIcon } from './play.svg';

interface DateControlsProps extends HTMLAttributes<HTMLElement> {
  disabled: boolean;
  datesList: string[];
  selectedDate?: string;
  onChangeDate: (e: string) => void;
  animate: boolean;
  onChangeAnimate: (e: boolean) => void;
}

const DateControls: React.SFC<DateControlsProps> = ({
  disabled,
  datesList,
  selectedDate,
  onChangeDate,
  animate,
  onChangeAnimate,
}) => {
  const handleChangeDate = (e: React.ChangeEvent<HTMLSelectElement>): void =>
    onChangeDate(e.target.value);
  const handleToggleAnimate = (): void => onChangeAnimate(!animate);

  return (
    <div className="legend-section">
      <label htmlFor="date" className="legend-section__date">
        Date
        <select
          id="date"
          value={selectedDate || ''}
          onChange={handleChangeDate}
          disabled={disabled || animate}
        >
          {datesList.map((date) => (
            <option key={`date-option-${date}`} value={date}>
              {date}
            </option>
          ))}
        </select>
      </label>
      <button
        id="animate-toggle-button"
        type="button"
        onClick={handleToggleAnimate}
        disabled={disabled}
      >
        {animate ? <PauseIcon className="animate-icon" /> : <PlayIcon className="animate-icon" />}
      </button>
    </div>
  );
};

export default DateControls;
