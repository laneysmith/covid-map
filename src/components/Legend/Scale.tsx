import React, { useMemo } from 'react';

import { formatNumber, chunkArray } from '../../utils';
import { Variable, ColorScales } from '../../types';

interface IScale {
  colorScales: ColorScales;
  selectedVariable: Variable;
}

const Scale: React.SFC<IScale> = ({ colorScales, selectedVariable }) => {
  const scale: (string | number)[] = colorScales[selectedVariable];
  const chunkedScale = useMemo(() => {
    const chunked = scale.length > 0 ? chunkArray(scale) : null;
    return chunked;
  }, [scale]);

  return (
    <div className="legend-section legend-section--column">
      <div className="legend-scale">
        {chunkedScale &&
          chunkedScale.map((step: (string | number)[], index: number) => {
            const [value, color] = step;
            return (
              <div
                key={`scale-step-${index}`}
                id={`scale-step-${index}`}
                className="legend-scale__step"
              >
                <span
                  className="legend-scale__step-swatch"
                  style={{
                    backgroundColor: `${color}`,
                    border: `1px solid rgba(0, 0, 0, 0.${index + 2})`,
                  }}
                ></span>{' '}
                {formatNumber(value)}
              </div>
            );
          })}
      </div>
      <div className="source-notes">
        Data from The New York Times
        <br />
        <a href="https://github.com/laneysmith/covid-map" target="_blank" rel="noopener noreferrer">
          View project source on github
        </a>
      </div>
    </div>
  );
};

export default Scale;
