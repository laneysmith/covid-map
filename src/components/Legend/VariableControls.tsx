import React from 'react';

import { Variable } from '../../types';

interface IVariableControls {
  disabled: boolean;
  selectedVariable: Variable;
  onChangeVariable: (value: Variable) => void;
}

const VariableControls: React.SFC<IVariableControls> = ({
  disabled,
  selectedVariable,
  onChangeVariable,
}) => {
  const handleChangeVariable = (e: React.ChangeEvent<HTMLInputElement>): void =>
    onChangeVariable(Variable[e.target.value as keyof typeof Variable]);

  return (
    <div className="legend-section">
      {Object.entries(Variable).map(([key, value]) => (
        <label key={value} htmlFor={`${value}-radio-button`}>
          <input
            type="radio"
            id={`${value}-radio-button`}
            name="variable"
            value={key}
            checked={selectedVariable === value}
            onChange={handleChangeVariable}
            disabled={disabled}
          />{' '}
          {`${value.charAt(0).toUpperCase()}${value.slice(1)}`}
        </label>
      ))}
    </div>
  );
};

export default VariableControls;
