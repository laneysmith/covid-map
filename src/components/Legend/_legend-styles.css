.legend {
  background-color: #fff;
  background-clip: padding-box;
  border: 2px solid rgba(0, 0, 0, 0.1);
  border-radius: 6px;
  bottom: 35px;
  position: absolute;
  left: 10px;
  z-index: 1;
}

.legend-section {
  display: flex;
  margin: 0;
  padding: 10px;
}

.legend-section:not(:last-child) {
  border-bottom: 1px solid #efefef;
}

.legend-section--column {
  flex-direction: column;
}

.legend label {
  font-weight: 600;
  margin-right: 8px;
}

.legend select {
  flex: 1;
  height: 20px;
  margin-right: 8px;
}

.legend-scale {
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-template-rows: repeat(5, 1fr);
  grid-column-gap: 5px;
  grid-row-gap: 0px;
  grid-auto-flow: column;
  padding: 0 0 5px 0;
}

.legend-scale__step {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.legend-scale__step-swatch {
  border-radius: 50%;
  display: inline-block;
  height: 12px;
  width: 12px;
  margin-right: 5px;
}

.source-notes {
  color: grey;
  font-size: 8pt;
  line-height: 10pt;
  max-width: 200px;
}

.source-notes a,
a:visited,
a:hover,
a:active {
  color: grey;
}

button {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 4px;
  border-radius: 3px;
  border: 1px solid #c0c0c0;
  background-color: white;
}

button:hover {
  cursor: pointer;
}

button:disabled {
  background-color: #f8f8f8;
}

button:disabled .animate-icon {
  fill: #C0C0C0;
}

select:focus,
button:focus {
  outline: none;
  border-color: #14899C;
}

.animate-icon {
  fill: #222222;
  height: 10px;
  width: 10px;
}