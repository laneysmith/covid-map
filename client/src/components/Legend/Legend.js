import React from 'react';

import DateControls from './DateControls';
import VariableControls from './VariableControls';
import Scale from './Scale';
import './_legend-styles.css';

const Legend = ({
  loaded,
  colorScales,
  allDates,
  animate,
  selectedVariable,
  selectedDate,
  onChangeAnimate,
  onChangeVariable,
  onChangeDate,
}) => (
  <div id="state-legend" className="legend">
    <DateControls
      disabled={!loaded}
      allDates={allDates}
      selectedDate={selectedDate}
      onChangeDate={onChangeDate}
      animate={animate} onChangeAnimate={onChangeAnimate} 
    />
    <VariableControls
      disabled={!loaded || animate}
      onChangeVariable={onChangeVariable}
      selectedVariable={selectedVariable}
    />
    <Scale colorScales={colorScales} selectedVariable={selectedVariable} />
  </div>
);

export default Legend;
