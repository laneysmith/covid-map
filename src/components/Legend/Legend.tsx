import React from 'react';

import { Variable, ColorScales } from '../../types';
import DateControls from './DateControls';
import VariableControls from './VariableControls';
import Scale from './Scale';
import './_legend-styles.css';

interface ILegend {
  loaded: boolean;
  colorScales: ColorScales;
  datesList: string[];
  selectedDate?: string;
  onChangeDate: (e: string) => void;
  animate: boolean;
  onChangeAnimate: (e: boolean) => void;
  selectedVariable: Variable;
  onChangeVariable: (value: Variable) => void;
}

const Legend: React.SFC<ILegend> = ({
  loaded,
  colorScales,
  datesList,
  animate,
  selectedVariable,
  selectedDate,
  onChangeAnimate,
  onChangeVariable,
  onChangeDate,
}) => (
  <div id="state-legend" className="legend">
    <DateControls
      disabled={!loaded}
      datesList={datesList}
      selectedDate={selectedDate}
      onChangeDate={onChangeDate}
      animate={animate}
      onChangeAnimate={onChangeAnimate}
    />
    <VariableControls
      disabled={!loaded || animate}
      onChangeVariable={onChangeVariable}
      selectedVariable={selectedVariable}
    />
    <Scale colorScales={colorScales} selectedVariable={selectedVariable} />
  </div>
);

export default Legend;
