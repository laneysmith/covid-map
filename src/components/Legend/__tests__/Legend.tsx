import React from 'react';
import { render } from '@testing-library/react';

import Legend from '../Legend';
import { Variable } from '../../../types';

describe('Legend', () => {
  const PROPS = {
    colorScales: {
      [Variable.CASES]: [
        10,
        '#E1F5FE',
        20,
        '#B3E5FC',
        30,
        '#81D4FA',
        40,
        '#4FC3F7',
        50,
        '#29B6F6',
        60,
        '#03A9F4',
        70,
        '#039BE5',
        80,
        '#0288D1',
        80,
        '#0277BD',
        100,
        '#01579B',
      ],
      [Variable.DEATHS]: [
        1,
        '#E1F5FE',
        2,
        '#B3E5FC',
        3,
        '#81D4FA',
        4,
        '#4FC3F7',
        5,
        '#29B6F6',
        6,
        '#03A9F4',
        7,
        '#039BE5',
        8,
        '#0288D1',
        8,
        '#0277BD',
        10,
        '#01579B',
      ],
    },
    datesList: ['2020-02-14', '2020-02-15', '2020-02-16'],
    selectedDate: '2020-02-14',
    selectedVariable: Variable.DEATHS,
    animate: false,
    onChangeDate: jest.fn(),
    onChangeVariable: jest.fn(),
    onChangeAnimate: jest.fn(),
  };

  it('calls renders correctly when loaded', () => {
    const { container } = render(<Legend {...PROPS} loaded />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('calls renders correctly when not loaded', () => {
    const { container } = render(<Legend {...PROPS} loaded={false} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
