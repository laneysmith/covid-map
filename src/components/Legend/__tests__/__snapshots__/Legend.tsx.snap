// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend calls renders correctly when loaded 1`] = `
<div
  class="legend"
  id="state-legend"
>
  <div
    class="legend-section"
  >
    <label
      for="date"
    >
      Date
      <select
        id="date"
      >
        <option
          value="2020-02-14"
        >
          2020-02-14
        </option>
        <option
          value="2020-02-15"
        >
          2020-02-15
        </option>
        <option
          value="2020-02-16"
        >
          2020-02-16
        </option>
      </select>
    </label>
    <button
      id="animate-toggle-button"
      type="button"
    >
      <svg
        class="animate-icon"
      >
        play.svg
      </svg>
    </button>
  </div>
  <div
    class="legend-section"
  >
    <label
      for="cases-radio-button"
    >
      <input
        id="cases-radio-button"
        name="variable"
        type="radio"
        value="CASES"
      />
       
      Cases
    </label>
    <label
      for="deaths-radio-button"
    >
      <input
        checked=""
        id="deaths-radio-button"
        name="variable"
        type="radio"
        value="DEATHS"
      />
       
      Deaths
    </label>
  </div>
  <div
    class="legend-section legend-section--column"
  >
    <div
      class="legend-scale"
    >
      <div
        class="legend-scale__step"
        id="scale-step-0"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(225, 245, 254);"
        />
        1
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-1"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(179, 229, 252);"
        />
        2
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-2"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(129, 212, 250);"
        />
        3
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-3"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(79, 195, 247);"
        />
        4
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-4"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(41, 182, 246);"
        />
        5
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-5"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(3, 169, 244);"
        />
        6
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-6"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(3, 155, 229);"
        />
        7
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-7"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(2, 136, 209);"
        />
        8
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-8"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(2, 119, 189);"
        />
        8
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-9"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(1, 87, 155);"
        />
        10
      </div>
    </div>
    <div
      class="source-notes"
    >
      Data from The New York Times
      <br />
      <a
        href="https://github.com/laneysmith/covid-map"
        rel="noopener noreferrer"
        target="_blank"
      >
        View project source on github
      </a>
    </div>
  </div>
</div>
`;

exports[`Legend calls renders correctly when not loaded 1`] = `
<div
  class="legend"
  id="state-legend"
>
  <div
    class="legend-section"
  >
    <label
      for="date"
    >
      Date
      <select
        disabled=""
        id="date"
      >
        <option
          value="2020-02-14"
        >
          2020-02-14
        </option>
        <option
          value="2020-02-15"
        >
          2020-02-15
        </option>
        <option
          value="2020-02-16"
        >
          2020-02-16
        </option>
      </select>
    </label>
    <button
      disabled=""
      id="animate-toggle-button"
      type="button"
    >
      <svg
        class="animate-icon"
      >
        play.svg
      </svg>
    </button>
  </div>
  <div
    class="legend-section"
  >
    <label
      for="cases-radio-button"
    >
      <input
        disabled=""
        id="cases-radio-button"
        name="variable"
        type="radio"
        value="CASES"
      />
       
      Cases
    </label>
    <label
      for="deaths-radio-button"
    >
      <input
        checked=""
        disabled=""
        id="deaths-radio-button"
        name="variable"
        type="radio"
        value="DEATHS"
      />
       
      Deaths
    </label>
  </div>
  <div
    class="legend-section legend-section--column"
  >
    <div
      class="legend-scale"
    >
      <div
        class="legend-scale__step"
        id="scale-step-0"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(225, 245, 254);"
        />
        1
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-1"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(179, 229, 252);"
        />
        2
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-2"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(129, 212, 250);"
        />
        3
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-3"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(79, 195, 247);"
        />
        4
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-4"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(41, 182, 246);"
        />
        5
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-5"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(3, 169, 244);"
        />
        6
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-6"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(3, 155, 229);"
        />
        7
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-7"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(2, 136, 209);"
        />
        8
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-8"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(2, 119, 189);"
        />
        8
      </div>
      <div
        class="legend-scale__step"
        id="scale-step-9"
      >
        <span
          class="legend-scale__step-swatch"
          style="background-color: rgb(1, 87, 155);"
        />
        10
      </div>
    </div>
    <div
      class="source-notes"
    >
      Data from The New York Times
      <br />
      <a
        href="https://github.com/laneysmith/covid-map"
        rel="noopener noreferrer"
        target="_blank"
      >
        View project source on github
      </a>
    </div>
  </div>
</div>
`;
